File: RxHttp/src/main/java/com/github/kongpf8848/rxhttp/converter/DownloadConverter.java
package com.github.kongpf8848.rxhttp.converter;

import com.github.kongpf8848.rxhttp.Platform;
import com.github.kongpf8848.rxhttp.ProgressResponseBody;
import com.github.kongpf8848.rxhttp.bean.DownloadInfo;
import com.github.kongpf8848.rxhttp.callback.DownloadCallback;
import com.github.kongpf8848.rxhttp.callback.ProgressCallback;
import com.github.kongpf8848.rxhttp.request.DownloadRequest;

import java.io.File;
import java.io.FileOutputStream;
import java.lang.reflect.Type;

import okhttp3.ResponseBody;
import okio.BufferedSink;
import okio.Okio;

public class DownloadConverter<T> implements IConverter<T> {

    private DownloadInfo downloadInfo;
    private DownloadCallback callback;
    private DownloadRequest downloadRequest;
    public DownloadConverter(DownloadRequest downloadRequest,DownloadCallback callback) {
        this.downloadRequest = downloadRequest;
        this.callback=callback;
        this.downloadInfo=new DownloadInfo(this.downloadRequest.getUrl(),this.downloadRequest.getDir(),this.downloadRequest.getFilename());
    }

    @Override
    public T convert(final ResponseBody body,Type type) throws Exception {
        try {
            File fileDir = new File(downloadInfo.getDestDir());
            if (!fileDir.exists()) {
                fileDir.mkdirs();
            }
            File file=new File(downloadInfo.getDestDir(),downloadInfo.getFileName());
            final long currentProgress;
            if(downloadRequest.isBreakpoint()){
                if(file.exists()){
                    currentProgress=file.length();
                }
                else{
                    currentProgress=0;
                }
            }
            else{
                currentProgress=0;
            }
            FileOutputStream fos=new FileOutputStream(file,downloadRequest.isBreakpoint());
            ProgressResponseBody progressResponseBody = new ProgressResponseBody(body, new ProgressCallback() {
                @Override
                public void onProgress(long readBytes,long totalBytes) {
                    downloadInfo.setProgress(currentProgress+readBytes);
                    Platform.get().defaultCallbackExecutor().execute(new Runnable() {
                        @Override
                        public void run() {
                            callback.onProgress(downloadInfo);
                        }
                    });
                }
            });
            downloadInfo.setTotal(currentProgress+progressResponseBody.contentLength());
            Platform.get().defaultCallbackExecutor().execute(new Runnable() {
                @Override
                public void run() {
                    callback.onProgress(downloadInfo);
                }
            });

            BufferedSink sink = Okio.buffer(Okio.sink(fos));
            sink.writeAll(progressResponseBody.source());
            sink.flush();
            sink.close();
            progressResponseBody.close();

        } catch (Exception e) {
            throw e;
        }
        return (T)downloadInfo;

    }
}

File: RxHttp/src/main/java/com/github/kongpf8848/rxhttp/converter/DownloadConverter.kt
package com.github.kongpf8848.rxhttp.converter

import com.github.kongpf8848.rxhttp.Platform
import com.github.kongpf8848.rxhttp.ProgressResponseBody
import com.github.kongpf8848.rxhttp.bean.DownloadInfo
import com.github.kongpf8848.rxhttp.callback.DownloadCallback
import com.github.kongpf8848.rxhttp.callback.ProgressCallback
import com.github.kongpf8848.rxhttp.request.DownloadRequest
import okhttp3.ResponseBody
import okio.buffer
import okio.sink
import java.io.File
import java.io.FileOutputStream
import java.lang.reflect.Type

class DownloadConverter<T>(private val downloadRequest: DownloadRequest, private val callback: DownloadCallback) : IConverter<T> {

    private val downloadInfo= DownloadInfo(downloadRequest.url, downloadRequest.dir, downloadRequest.filename)

    @Throws(Exception::class)
    override fun convert(body: ResponseBody, type: Type): T {
        try {
            val fileDir = File(downloadInfo.dir)
            if (!fileDir.exists()) {
                fileDir.mkdirs()
            }
            val file = File(downloadInfo.dir, downloadInfo.fileName)
            val currentProgress= if (downloadRequest.isBreakpoint) {
                if (file.exists()) {
                    file.length()
                } else {
                    0L
                }
            } else {
                0L
            }
            val fos = FileOutputStream(file, downloadRequest.isBreakpoint)
            val progressResponseBody = ProgressResponseBody(body, object : ProgressCallback {
                override fun onProgress(readBytes: Long, totalBytes: Long) {
                    downloadInfo.progress=currentProgress+readBytes
                    Platform.get().defaultCallbackExecutor()!!.execute { callback.onProgress(downloadInfo.progress, currentProgress+totalBytes) }
                }
            })
            downloadInfo.total = currentProgress + progressResponseBody.contentLength()
            Platform.get().defaultCallbackExecutor()?.execute { callback.onProgress(currentProgress,downloadInfo.total) }
            val sink = fos.sink().buffer()
            sink.writeAll(progressResponseBody.source())
            sink.flush()
            sink.close()
            progressResponseBody.close()
        } catch (e: Exception) {
            throw e
        }
        return downloadInfo as T
    }

}